ENH: Allow toggling madvise hugepage and fix default

seberg opened the pull request:
By default this disables madvise hugepage on kernels before 4.6, since
we expect that these typically see large performance regressions when
using hugepages due to slow defragementation code presumably fixed by:

https://github.com/torvalds/linux/commit/7cf91a98e607c2f935dbcc177d70011e95b8faff

This adds support to set the behaviour at startup time through the
``NUMPY_MADVISE_HUGEPAGE`` environment variable.

Fixes gh-15545

----

Still needs some documentation somewhere probably... I am not sure where though, a new site listing all environment variables used during compile or startup time? Relaxed strides, experimental array function protocol, this one, ...?

seberg committed:
ENH: Allow toggling madvise hugepage and fix default

By default this disables madvise hugepage on kernels before 4.6, since
we expect that these typically see large performance regressions when
using hugepages due to slow defragementation code presumably fixed by:

https://github.com/torvalds/linux/commit/7cf91a98e607c2f935dbcc177d70011e95b8faff

This adds support to set the behaviour at startup time through the
``NUMPY_MADVISE_HUGEPAGE`` environment variable.

Fixes gh-15545

mattip commented:
Maybe a page about "Global State", and mention `threadpoolctl` as well, kind of like in the [python documentation](https://docs.python.org/3/using/cmdline.html#environment-variables)

seberg committed:
DOC: Note few global global options we currently have in refdocs

These are options that are controlled typically through environment
variable at startup or compile time.

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+-----------------------------------------
+
+NumPy itself is normally intentionally limited to a single thread
+during function calls, however it does support multiple python
```
```suggestion
during function calls, however it does support multiple Python
```

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+
+When working with very large arrays on modern Linux kernels,
+you can experience a significant speedup when transparent
+hugepage is enabled.
```
Do you want to include an external link to linux docs on transparent hugepages? If so, you could consider this: https://www.kernel.org/doc/html/latest/vm/transhuge.html

seberg replied:
good idea.

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+you can experience a significant speedup when transparent
+hugepage is enabled.
+This may always be the case or may use ``madvise`` option as
+seen by::
```
```suggestion
The current system policy for transparent hugepages can be 
seen by::
```

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+    cat /sys/kernel/mm/transparent_hugepage/enabled
+
+on most kernels.  When set to ``madvise`` NumPy will typically
+use enable hugepages for a performance boost. This behaviour can
```
```suggestion
use hugepages for a performance boost. This behaviour can
```

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+This flag is checked at import time.
+
+
+Interpoerabilty
```
```suggestion
Interoperability
```
Could even say "Interoperability-related Options to be consistent with "Performance-related Options"

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+
+The array function protocol which allows array-like objects to
+hook into the NumPy API is currently enabled by default.
+It can be disabled using::
```
Perhaps mention that the feature was introduced in v1.17 and is enabled by default

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+
+    NUMPY_EXPERIMENTAL_ARRAY_FUNCTION=0
+
+See also `class.__array_function__` for more information.
```
```suggestion
See also :py:meth:`numpy.class.__array_function__` for more information.
```
I think that will make the link work

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+This flag is checked at import time.
+
+
+Debugging
```
Same suggestion as above - maybe "Debugging-related Options" to be consistent

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+Relaxed Strides Checking
+------------------------
+
+The *compile* time environment variables::
```
```suggestion
The *compile-time* environment variables::
```

rossbar commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+in C which iterates through arrays that may or may not be
+contiguous in memory.
+Most users will have no reason to change these, for details
+please see the `memory layout <memory-layout>`_ documentation.
```
```suggestion
please see the :ref:`memory layout <memory-layout>`_ documentation.
```
I think this needs an explicit ref role

rossbar reviewed (changes requested):
This seems like a nice feature - I mostly did a once-over on the docs.

Just for kicks, I tried this on my system (kernel v5.5.9) and did indeed see that performance was generally better with NumPy's use of hugepages was enabled. Using the original examples from #15545:

```python
>>> import numpy as np; print(np.use_hugepage)
1
>>> n = int(1e9)
>>> %timeit np.zeros(n)                                                                  
5.52 µs ± 75.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit np.random.rand(n)                                                            
5.87 s ± 76.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit np.linspace(0, 100, n)                                                       
2.49 s ± 50.8 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit np.exp(np.zeros(n))                                                          
7.05 s ± 152 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```
And with `NUMPY_MADVISE_HUGEPAGE=0`:
```python
>>> import numpy as np; print(np.use_hugepage)
0
>>> n = int(1e9)
>>> %timeit np.zeros(n)                                                                  
7.86 µs ± 30.2 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
>>> %timeit np.random.rand(n)                                                            
6.97 s ± 77 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit np.linspace(0, 100, n)                                                         
3.78 s ± 12.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
>>> %timeit np.exp(np.zeros(n))                                                          
8.09 s ± 68.4 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

seberg commented:
Meh, I confirmed that kernel 4.8 is fixed, then I went down a wrong turn and lost my progress on testing 4.6 as well :(.

seberg commented on doc/source/reference/global_state.rst:
```diff
@@ -0,0 +1,85 @@
+Global State
+************
+
+NumPy has a few startup time, compile, or runtime options
```
Yeah, import time.

seberg committed:
Apply suggestions from code review by Ross

Co-Authored-By: Ross Barnowski <rossbar@berkeley.edu>

seberg committed:
Further DOC-Fixup base don Ross' review

seberg commented:
Not sure if it makes sense, but we could consider backporting this, although maybe it is too high risk? Since it tries to at least mitigate a performance regression.

anirudh2290 commented on numpy/__init__.py:
```diff
@@ -286,3 +286,24 @@ def _mac_os_check():
+        kernel_version = os.uname().release.split(".")[:2]
+        kernel_version = tuple(int(v) for v in kernel_version)
+        if kernel_version < (4, 6):
+            use_hugepage = 0
```
I dont know if this will help here, but is it worth putting a message, saying that if using version less than 4.6 and you notice issues with large arrays try setting NUMPY_MADVISE_HUGEPAGE=1. If we do this we can also backport and it will be safer since there is a message shown to the user to fix the issue.

seberg replied:
Well, there seemed to be some consensus around not trying to guess what is probably right for most people and instead adding into an FAQ style tip page somewhere? I actually like guessing if it helps 90% of the people, since I think very few will find the setting or even know they are getting terrible performance...

mattip replied:
I think guessing is fine. Maybe the new troubleshooting page would be a good page to reference this setting.

mattip replied:
Ahh, I see you have already added a global_state page. That seems fine.

seberg commented on numpy/core/src/multiarray/alloc.c:
```diff
@@ -47,6 +47,25 @@ typedef struct {
+    }
+    _madvise_hugepage = enabled;
+    if (was_enabled) {
+        Py_RETURN_TRUE;
```
Not sure I follow, this function sets the behaviour and returns the current state, either `True` or `False` in the line below?

anirudh2290 replied:
but `was_enabled` was set to `_madvise_hugepage` value which is set to 1 above, it doesnt seem to change. isnt `was_enabled` always 1 ?

seberg replied:
Maybe I have to add a test to proof that it does, I admit. `was_enabled` is a *copy* of the global static `_madvise_hugepage`, which is modified with `_madvise_hugepage = enabled`. So if `enabled` is 0, on the next call `was_enabled` will be 0.

anirudh2290 replied:
okay thanks for explaining, i guess i misunderstood this function completely :grin: . i was expecting it to return true for platforms where _set_madvise_hugepage works (i.e. it is able to use this env variable  and false where it doesnt, didnt realize was_enabled was for the next function call).

mattip replied:
A comment before the function with an explanation may help prevent future confusion.

mattip commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -4159,6 +4161,8 @@ static struct PyMethodDef array_module_methods[] = {
     {"_add_newdoc_ufunc", (PyCFunction)add_newdoc_ufunc,
         METH_VARARGS, NULL},
+    {"_set_madvise_hugepage", (PyCFunction)_set_madvise_hugepage,
+        METH_O, "Toggle and return madvise hugepage (no OS support check)."},
```
maybe extend this docstring, even though it is a private function:
```
_set_madvise_hugepage(tf: bool) -> bool
Set  or unset use of ``madvise (2)`` MADV_HUGEPAGE support when allocating
the array data. Returns the previously set value. See `global_state` for more
information.
```

seberg committed:
Fixup: Implement suggestions by Matti

pentschev reviewed (approved):
The proposed solution looks good to me, thanks for doing the work @seberg !

mattip commented:
close/reopen to restart builds

mattip commented:
Merging, the failing s390x job is unrelated.

mattip commented:
Thanks @seberg

bmerry commented on doc/release/upcoming_changes/15769.improvement.rst:
```diff
@@ -0,0 +1,15 @@
+
+On Linux NumPy has previously added support for madavise
+hugepages which can improve performance for very large arrays.
+Unfortunately, on older Kernel versions this led to peformance
```
I realise I'm late to the party since this has already been merged, but "peformance" -> "performance".

Megaland90 commented:
hi i have probleme with kernel name ovh "4.19-ovh-xxxx-std-ipv6-64" :

kernel_version = tuple(int(v) for v in kernel_version)
ValueError: invalid literal for int() with base 10: '19-ovh-xxxx-std-ipv6-64'

mattip commented:
@Megaland90 You are looking for gh-16679 which will be part of the upcoming 1.19.1.

Megaland90 commented:
Thanks @mattip  :D

mattip commented:
@Megaland90 thanks for the report. For future reference, opening a new issue rather than adding comments to a PR helps keep things more organized (although I just added two more comments myself :))
